File: _may_messenger_web_client/src/services/typingIndicatorService.ts
/**
 * Typing Indicator Service - управление индикаторами набора текста с debouncing
 * Предотвращает спам на сервер событиями набора текста
 */

type SendTypingIndicator = (chatId: string, isTyping: boolean) => Promise<void>;

class TypingIndicatorService {
  private readonly sendTypingIndicator: SendTypingIndicator;

  // Конфигурация debouncing
  private static readonly TYPING_DEBOUNCE_DELAY = 300; // 300мс
  private static readonly TYPING_STOP_DELAY = 3000; // 3 секунды

  // Отслеживание состояния
  private debounceTimers = new Map<string, number>();
  private stopTimers = new Map<string, number>();
  private currentTypingStates = new Map<string, boolean>();

  constructor(sendTypingIndicator: SendTypingIndicator) {
    this.sendTypingIndicator = sendTypingIndicator;
  }

  /**
   * Пользователь начал набирать текст в чате
   * Debounce событие, чтобы избежать спама
   */
  onTyping(chatId: string): void {
    // Отменить существующий таймер debounce
    const existingTimer = this.debounceTimers.get(chatId);
    if (existingTimer) {
      clearTimeout(existingTimer);
    }

    // Установить новый таймер debounce
    const timer = window.setTimeout(() => {
      this.setTyping(chatId, true);
    }, TypingIndicatorService.TYPING_DEBOUNCE_DELAY);

    this.debounceTimers.set(chatId, timer);

    // Сбросить таймер автоостановки (пользователь все еще печатает)
    this.resetStopTimer(chatId);
  }

  /**
   * Пользователь перестал набирать текст в чате (явно)
   */
  onStoppedTyping(chatId: string): void {
    this.setTyping(chatId, false);

    const debounceTimer = this.debounceTimers.get(chatId);
    if (debounceTimer) {
      clearTimeout(debounceTimer);
      this.debounceTimers.delete(chatId);
    }

    const stopTimer = this.stopTimers.get(chatId);
    if (stopTimer) {
      clearTimeout(stopTimer);
      this.stopTimers.delete(chatId);
    }
  }

  /**
   * Внутренний метод для установки состояния набора текста
   */
  private setTyping(chatId: string, isTyping: boolean): void {
    // Отправлять только если состояние изменилось
    if (this.currentTypingStates.get(chatId) === isTyping) {
      return;
    }

    this.currentTypingStates.set(chatId, isTyping);

    try {
      this.sendTypingIndicator(chatId, isTyping);
      console.log(`[TypingIndicator] Sent typing=${isTyping} for chat ${chatId}`);
    } catch (error) {
      console.error('[TypingIndicator] Error sending typing indicator:', error);
    }

    // Если начали печатать, установить таймер автоостановки
    if (isTyping) {
      this.resetStopTimer(chatId);
    }
  }

  /**
   * Сбросить таймер автоостановки
   */
  private resetStopTimer(chatId: string): void {
    const existingTimer = this.stopTimers.get(chatId);
    if (existingTimer) {
      clearTimeout(existingTimer);
    }

    const timer = window.setTimeout(() => {
      this.setTyping(chatId, false);
    }, TypingIndicatorService.TYPING_STOP_DELAY);

    this.stopTimers.set(chatId, timer);
  }

  /**
   * Очистить таймеры для конкретного чата
   */
  cleanupChat(chatId: string): void {
    const debounceTimer = this.debounceTimers.get(chatId);
    if (debounceTimer) {
      clearTimeout(debounceTimer);
      this.debounceTimers.delete(chatId);
    }

    const stopTimer = this.stopTimers.get(chatId);
    if (stopTimer) {
      clearTimeout(stopTimer);
      this.stopTimers.delete(chatId);
    }

    this.currentTypingStates.delete(chatId);
    console.log(`[TypingIndicator] Cleaned up chat ${chatId}`);
  }

  /**
   * Освободить все ресурсы
   */
  dispose(): void {
    this.debounceTimers.forEach((timer) => clearTimeout(timer));
    this.stopTimers.forEach((timer) => clearTimeout(timer));

    this.debounceTimers.clear();
    this.stopTimers.clear();
    this.currentTypingStates.clear();

    console.log('[TypingIndicator] Disposed');
  }

  /**
   * Получить статистику для отладки
   */
  getStats(): {
    activeChats: number;
    typingChats: string[];
  } {
    const typingChats = Array.from(this.currentTypingStates.entries())
      .filter(([, isTyping]) => isTyping)
      .map(([chatId]) => chatId);

    return {
      activeChats: this.currentTypingStates.size,
      typingChats,
    };
  }
}

export default TypingIndicatorService;

File: _may_messenger_mobile_app/lib/data/services/typing_indicator_service.dart
import 'dart:async';
import 'package:logger/logger.dart';

/// Service for managing typing indicators with debouncing
/// Prevents spamming the server with typing events
class TypingIndicatorService {
  final Logger _logger = Logger();
  final Function(String chatId, bool isTyping) _sendTypingIndicator;
  
  // Debouncing configuration
  static const Duration typingDebounceDelay = Duration(milliseconds: 300);
  static const Duration typingStopDelay = Duration(seconds: 3);
  
  // State tracking
  final Map<String, Timer?> _debounceTimers = {};
  final Map<String, Timer?> _stopTimers = {};
  final Map<String, bool> _currentTypingStates = {};

  TypingIndicatorService(this._sendTypingIndicator);

  /// User started typing in a chat
  /// Debounces the event to avoid spamming
  void onTyping(String chatId) {
    // Cancel existing debounce timer
    _debounceTimers[chatId]?.cancel();
    
    // Set new debounce timer
    _debounceTimers[chatId] = Timer(typingDebounceDelay, () {
      _setTyping(chatId, true);
    });
    
    // Reset stop timer (user is still typing)
    _resetStopTimer(chatId);
  }

  /// User stopped typing in a chat (explicit)
  void onStoppedTyping(String chatId) {
    _setTyping(chatId, false);
    _debounceTimers[chatId]?.cancel();
    _stopTimers[chatId]?.cancel();
  }

  /// Internal method to set typing state
  void _setTyping(String chatId, bool isTyping) {
    // Only send if state changed
    if (_currentTypingStates[chatId] == isTyping) {
      return;
    }

    _currentTypingStates[chatId] = isTyping;
    
    try {
      _sendTypingIndicator(chatId, isTyping);
      _logger.d('[TypingIndicator] Sent typing=$isTyping for chat $chatId');
    } catch (e) {
      _logger.e('[TypingIndicator] Error sending typing indicator', error: e);
    }

    // If started typing, set auto-stop timer
    if (isTyping) {
      _resetStopTimer(chatId);
    }
  }

  /// Reset the auto-stop timer
  void _resetStopTimer(String chatId) {
    _stopTimers[chatId]?.cancel();
    
    _stopTimers[chatId] = Timer(typingStopDelay, () {
      _setTyping(chatId, false);
    });
  }

  /// Clean up timers for a specific chat
  void cleanupChat(String chatId) {
    _debounceTimers[chatId]?.cancel();
    _stopTimers[chatId]?.cancel();
    _debounceTimers.remove(chatId);
    _stopTimers.remove(chatId);
    _currentTypingStates.remove(chatId);
    _logger.d('[TypingIndicator] Cleaned up chat $chatId');
  }

  /// Dispose all resources
  void dispose() {
    for (final timer in _debounceTimers.values) {
      timer?.cancel();
    }
    for (final timer in _stopTimers.values) {
      timer?.cancel();
    }
    _debounceTimers.clear();
    _stopTimers.clear();
    _currentTypingStates.clear();
    _logger.d('[TypingIndicator] Disposed');
  }

  /// Get statistics for debugging
  Map<String, dynamic> getStats() {
    return {
      'activeChats': _currentTypingStates.length,
      'typingChats': _currentTypingStates.entries
          .where((e) => e.value)
          .map((e) => e.key)
          .toList(),
    };
  }
}

